File: backend/F1Fantasy/Controllers/PitStopController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using F1Fantasy.Models;
using F1Fantasy.Services;

namespace F1Fantasy.Controllers;

[Authorize]
[ApiController]
[Route("api/[controller]")]
public class PitStopController : ControllerBase
{
    private readonly PitStopService _pitStopService;
    private readonly ILogger<PitStopController> _logger;

    public PitStopController(PitStopService pitStopService, ILogger<PitStopController> logger)
    {
        _pitStopService = pitStopService;
        _logger = logger;
    }

    [HttpGet("season/{season}/round/{round}")]
    public async Task<ActionResult<RaceWithPitStops>> GetPitStopsByRace(string season, string round)
    {
        try
        {
            _logger.LogInformation("GET /api/pitstop/season/{Season}/round/{Round} - Fetching pit stops for race", 
                season, round);
            var pitStops = await _pitStopService.GetPitStopsByRaceAsync(season, round);
            
            if (pitStops == null)
            {
                _logger.LogWarning("No pit stops found for season {Season}, round {Round}", season, round);
                return NotFound(new { message = $"No pit stops found for season {season}, round {round}" });
            }
            
            _logger.LogInformation("Successfully retrieved {Count} pit stops for season {Season}, round {Round}", 
                pitStops.PitStops?.Count ?? 0, season, round);
            return Ok(pitStops);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching pit stops for season {Season}, round {Round}", season, round);
            throw;
        }
    }

    [HttpGet("season/{season}/round/{round}/driver/{driverId}")]
    public async Task<ActionResult<IEnumerable<PitStop>>> GetPitStopsByDriver(string season, string round, string driverId)
    {
        try
        {
            _logger.LogInformation("GET /api/pitstop/season/{Season}/round/{Round}/driver/{DriverId} - Fetching driver pit stops", 
                season, round, driverId);
            var pitStops = await _pitStopService.GetPitStopsByDriverAsync(season, round, driverId);
            
            if (!pitStops.Any())
            {
                _logger.LogWarning("No pit stops found for season {Season}, round {Round}, driver {DriverId}", 
                    season, round, driverId);
                return NotFound(new { message = $"No pit stops found for driver {driverId} in season {season}, round {round}" });
            }
            
            _logger.LogInformation("Successfully retrieved {Count} pit stops for driver {DriverId}", 
                pitStops.Count(), driverId);
            return Ok(pitStops);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching pit stops for season {Season}, round {Round}, driver {DriverId}", 
                season, round, driverId);
            throw;
        }
    }

    [HttpGet("cached")]
    public async Task<ActionResult<IEnumerable<PitStop>>> GetCachedPitStops()
    {
        try
        {
            _logger.LogInformation("GET /api/pitstop/cached - Fetching cached pit stops");
            var pitStops = await _pitStopService.GetCachedPitStopsAsync();
            
            if (!pitStops.Any())
            {
                _logger.LogWarning("No cached pit stops found. Database may be empty.");
                return Ok(new { 
                    message = "No cached pit stops found. Try calling /api/pitstop/season/2025/round/1 first to populate the cache.",
                    pitStops = pitStops 
                });
            }
            
            _logger.LogInformation("Successfully retrieved {Count} cached pit stops", pitStops.Count());
            return Ok(pitStops);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching cached pit stops");
            throw;
        }
    }
}

File: backendJava/f1fantasy/src/main/java/no/f1fantasy/controller/PitStopController.java
package no.f1fantasy.controller;

import java.util.List;
import java.util.HashMap;
import java.util.Map;

import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import no.f1fantasy.entity.PitStop;
import no.f1fantasy.service.PitStopService;
import no.f1fantasy.service.RaceWithPitStops;

@RestController
@RequestMapping("/api/pitstop")
@PreAuthorize("isAuthenticated()")
@SuppressWarnings("null")
public class PitStopController {

    private final PitStopService pitStopService;

    public PitStopController(PitStopService pitStopService) {
        this.pitStopService = pitStopService;
    }

    @GetMapping("/season/{season}/round/{round}")
    public ResponseEntity<?> getPitStopsByRace(@PathVariable String season, @PathVariable String round) {
        try {
            java.util.Optional<RaceWithPitStops> pitStops = pitStopService.getPitStopsByRace(season, round);
            if (!pitStops.isPresent()) {
                return ResponseEntity.status(404).body(errorMap("No pit stops found for season " + season + ", round " + round));
            }
            return ResponseEntity.ok(pitStops.get());
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching pit stops for season " + season + ", round " + round));
        }
    }

    @GetMapping("/season/{season}/round/{round}/driver/{driverId}")
    public ResponseEntity<?> getPitStopsByDriver(@PathVariable String season, @PathVariable String round, @PathVariable String driverId) {
        try {
            List<PitStop> pitStops = pitStopService.getPitStopsByDriver(season, round, driverId);
            if (pitStops == null || pitStops.isEmpty()) {
                return ResponseEntity.status(404).body(errorMap("No pit stops found for driver " + driverId + " in season " + season + ", round " + round));
            }
            return ResponseEntity.ok(pitStops);
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching pit stops for driver " + driverId));
        }
    }

    @GetMapping("/cached")
    public ResponseEntity<?> getCachedPitStops() {
        try {
            List<PitStop> pitStops = pitStopService.getCachedPitStops();
            if (pitStops == null || pitStops.isEmpty()) {
                Map<String, Object> response = new HashMap<>();
                response.put("message", "No cached pit stops found");
                response.put("pitStops", pitStops);
                return ResponseEntity.ok(response);
            }
            return ResponseEntity.ok(pitStops);
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching cached pit stops"));
        }
    }

    private Map<String, String> errorMap(String message) {
        Map<String, String> error = new HashMap<>();
        error.put("error", message);
        return error;
    }
}
